tsl/src/compression/algorithms/dictionary.c

```diff
@@ -123,6 +125,26 @@ dictionary_compressor_append_null_value(Compressor *compressor)
 	dictionary_compressor_append_null(extended->internal);
 }
 
+static bool
+dictionary_compressor_is_full(Compressor *compressor, Datum val)
+{
+	ExtendedCompressor *extended = (ExtendedCompressor *) compressor;
+	if (extended->internal == NULL)
+		extended->internal = dictionary_compressor_alloc(extended->element_type);
+
+	Size datum_size_and_align;
+	DictionaryCompressor *dict_comp = (DictionaryCompressor *) extended->internal;
+	if (datum_serializer_value_may_be_toasted(dict_comp->serializer))
+		val = PointerGetDatum(PG_DETOAST_DATUM_PACKED(val));
```

Comment:
This can potentially lead to double detoasting (here and then in the actual compression) which will be a major performance problem. Let's detoast in the caller and assert here.